### PR TITLE
sql: support BETWEEN SYMMETRIC

### PR DESCRIPTION
**Pull request by benesch**
BETWEEN SYMMETRIC is a SQL99 feature supported by Postgres that allows
arguments to be specified in non-sorted order. While a BETWEEN b AND c
means:

    b <= a <= c

a BETWEEN SYMMETRIC b AND c means:

    (b <= a <= c) OR (c <= a <= b)

Concretely, 2 BETWEEN 3 AND 1 is false, but 2 BETWEEN SYMMETRIC 3 AND 1
is true.

Previously, we would accept the BETWEEN SYMMETRIC syntax but ignored the
SYMMETRIC modifier, treating it just like BETWEEN.

Release note (bug fix): Fix BETWEEN SYMMETRIC, which was incorrectly
considered an alias for BETWEEN. Per the SQL99 specification, BETWEEN
SYMMETRIC is like BETWEEN, except that its arguments are automatically
swapped if they would specify an empty range.

### CODE REVIEW

**Review by jordanlewis — COMMENTED** [review]
Any particular reason you're adding this? Partitioning related I assume? I'm curious where you guys landed with the partitioning syntax.

Also hmm, it looks like we already supported the syntax but just returned incorrect results? If that's the case the release note needs updating.

BTW, I had to do the following rearrangement of the commit message to figure out what it meant, in case this helps anyone else:

```
BETWEEN ASYMMETRIC: b <= a < = c
BETWEEN SYMMETRIC:  b <= a <= c OR c <= a <= b
```

File: pkg/sql/sem/tree/eval_test.go [hunk 1]
```diff
@@ -516,9 +516,15 @@ func TestEval(t *testing.T) {
...
 		// Range conditions.
+		{`0 BETWEEN 1 AND 3`, `false`},
+		{`1 BETWEEN 1 AND 3`, `true`},
 		{`2 BETWEEN 1 AND 3`, `true`},
+		{`3 BETWEEN 1 AND 3`, `true`},
+		{`4 BETWEEN 1 AND 3`, `false`},
 		{`1 NOT BETWEEN 2 AND 3`, `true`},
 		{`'foo' BETWEEN 'a' AND 'z'`, `true`},
+		{`2 BETWEEN 3 AND 1`, `false`},
+		{`2 BETWEEN SYMMETRIC 3 AND 1`, `true`},
```
Comment (jordanlewis) [inline pkg/sql/sem/tree/eval_test.go]: I'd like to see a few more tests. What about a test where the from and to are swapped? NOT BETWEEN SYMMETRIC? BETWEEN ASYMMETRIC? etc.

**Comment by benesch** [comment]
Partitioning related only in that I noticed we supported the syntax but failed to do the right thing.

**Comment by benesch** [comment]
Anyway, updated with a lot more tests. PTAL!

**Review by jordanlewis — APPROVED** [review]
LGTM!

**Comment by benesch** [comment]
TFTR!

**Pull request merged**